components/metrics/lib.rs

```diff
@@ -1,77 +1,278 @@
 extern crate servo_config;
+extern crate time;
 
 use gfx::display_list::{DisplayItem, DisplayList};
 use gfx_traits::Epoch;
 use ipc_channel::ipc::IpcSender;
 use msg::constellation_msg::PipelineId;
 use profile_traits::time::{ProfilerChan, ProfilerCategory, send_profile_data};
 use profile_traits::time::TimerMetadata;
-use script_traits::{ConstellationControlMsg, LayoutMsg, PaintMetricType};
+use script_traits::{ConstellationControlMsg, LayoutMsg, PWMType};
 use servo_config::opts;
 use std::cell::{Cell, RefCell};
+use std::cmp::Ordering;
 use std::collections::HashMap;
+use std::time::{Instant, Duration};
+use time::precise_time_ns;
 
 pub trait ProfilerMetadataFactory {
     fn new_metadata(&self) -> Option<TimerMetadata>;
 }
 
-macro_rules! make_time_setter(
-    ( $attr:ident, $func:ident, $category:ident, $label:expr, $metric_type:path ) => (
-        fn $func(&self,
-                 profiler_metadata: Option<TimerMetadata>,
-                 paint_time: f64) {
-            if self.$attr.get().is_some() {
-                return;
-            }
+pub trait ProgressiveWebMetric {
+    fn get_navigation_start(&self) -> Option<f64>;
+    fn set_navigation_start(&mut self, time: f64);
+    fn get_time_profiler_chan(&self) -> &ProfilerChan;
+    fn send_queued_constellation_msg(&self, name: PWMType, time: f64);
+}
+
+lazy_static!{
+    static ref TEN_SECONDS: Duration = Duration::new(10, 0);
+    pub static ref MAX_TASK_TIME: u64 = 50000000;
+}
+
+
+// there should be some requirements on self
+// can i have self<U> ?
+pub fn set_metric<U: ProgressiveWebMetric>(
+    pwm: &U,
+    metadata: Option<TimerMetadata>,
+    metric_type: PWMType,
+    category: ProfilerCategory,
+    attr: &Cell<Option<f64>>,
+    metric_time: Option<f64>,
+) {
+    let navigation_start = match pwm.get_navigation_start() {
+        Some(time) => time,
+        None => {
+            warn!("Trying to set metric before navigation start");
+            return;
+        }
+    };
+
+    let now = match metric_time {
+        Some(time) => time,
+        None => precise_time_ns() as f64,
+    };
+    let time = now - navigation_start;
+    attr.set(Some(time));
+
+    // Queue performance observer notification.
+    pwm.send_queued_constellation_msg(metric_type, time);
+
+    // Send the metric to the time profiler.
+    send_profile_data(
+        category,
+        metadata,
+        &pwm.get_time_profiler_chan(),
+        time as u64,
+        time as u64,
+        0,
+        0,
+    );
+
+    // Print the metric to console if the print-pwm option was given.
+    if opts::get().print_pwm {
+        println!("{:?} {:?}", metric_type, time);
+    }
+
+}
+
+// https://github.com/GoogleChrome/lighthouse/issues/27
+#[derive(MallocSizeOf)]
+pub struct InteractiveMetrics {
+    navigation_start: Cell<Option<f64>>,
+    dom_content_loaded: Cell<Option<f64>>,
+    main_thread_available: Cell<Option<f64>>,
+    time_to_interactive: Cell<Option<f64>>,
```

Comment:
I would be useful to document what these values mean. In particular, I'm not sure what main_thread_available represents.